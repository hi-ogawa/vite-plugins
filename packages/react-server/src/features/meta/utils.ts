// cf. https://github.com/vercel/next.js/blob/4b7924b15593322633fe2847f52bac8dbd5d9047/packages/next/src/lib/metadata/types/metadata-interface.ts#L40
export type Metadata = {
  title?: null | string;
  description?: null | string;
};
