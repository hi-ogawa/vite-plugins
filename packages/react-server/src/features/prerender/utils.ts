export type PPRData = {
  preludeString: string;
  postponed: any;
};

export type PPRManifest = {
  entries: Record<string, PPRData>;
};

export async function streamToString(stream: ReadableStream<Uint8Array>) {
  let s = "";
  await stream.pipeThrough(new TextDecoderStream()).pipeTo(
    new WritableStream({
      write(c) {
        s += c;
      },
    }),
  );
  return s;
}
