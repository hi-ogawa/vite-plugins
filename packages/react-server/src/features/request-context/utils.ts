// we don't require async hooks and fallbacks to sync context
// (see virtual:inject-async-local-storage)

type ContextStorage<T> = {
  run<R>(store: T, callback: () => R): R;
  getStore(): T | undefined;
};

export function createContextStorage<T>(): ContextStorage<T> {
  if ((globalThis as any).AsyncLocalStorage) {
    return new (globalThis as any).AsyncLocalStorage();
  }
  return createSyncStorage();
}

function createSyncStorage<T>(): ContextStorage<T> {
  let current: T | undefined;
  return {
    run(store, callback) {
      current = store;
      const result = callback();
      current = undefined;
      return result;
    },
    getStore() {
      return current;
    },
  };
}
