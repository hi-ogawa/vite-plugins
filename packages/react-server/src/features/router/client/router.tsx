import {
  type HistoryLocation,
  type RouterHistory,
  createBrowserHistory,
} from "@tanstack/history";
import React from "react";
import { TinyStore, useStore } from "./store-utils";

type RouterState = {
  history: Omit<RouterHistory, "location">; // hide location from API since history is mutable
  location: HistoryLocation;
  isPending: boolean;
  isActionPending: boolean;
};

export class Router {
  public store: TinyStore<RouterState>;

  constructor(private history: RouterHistory) {
    this.store = new TinyStore<RouterState>({
      history,
      location: history.location,
      isPending: false,
      isActionPending: false,
    });
  }

  setup() {
    return this.history.subscribe(() => {
      this.store.set((s) => ({
        ...s,
        location: this.history.location,
      }));
    });
  }
}

export const RouterContext = React.createContext<Router>(undefined!);

export function useRouter<U = RouterState>(select?: (v: RouterState) => U) {
  const router = React.useContext(RouterContext);
  return useStore(router.store, select);
}

export function createEncodedBrowserHistory() {
  // patch push/replace so that location object consistently includes encoded url
  // (i.e. history.push("/âœ…") should set { pathname: "/%E2%9C%85" } as state)
  // cf.
  // https://github.com/remix-run/react-router/pull/9477
  // https://github.com/TanStack/router/issues/1441

  const history = createBrowserHistory();

  function encode(href: string) {
    const url = new URL(href, window.location.origin);
    return url.href.slice(url.origin.length);
  }

  function wrapEncode(f: typeof history.push): typeof history.push {
    return (path, state) => f(encode(path), state);
  }

  history.push = wrapEncode(history.push);
  history.replace = wrapEncode(history.replace);
  return history;
}
