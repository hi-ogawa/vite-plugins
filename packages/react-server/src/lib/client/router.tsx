import type { HistoryLocation, RouterHistory } from "@tanstack/history";
import React from "react";
import { TinyStore, useStore } from "./store-utils";

type RouterState = {
  history: Omit<RouterHistory, "location">; // hide location from API since history is mutable
  location: HistoryLocation;
  isPending: boolean;
  isActionPending: boolean;
};

export class Router {
  public store: TinyStore<RouterState>;

  constructor(public history: RouterHistory) {
    this.store = new TinyStore<RouterState>({
      history,
      location: history.location,
      isPending: false,
      isActionPending: false,
    });
  }

  setup() {
    return this.history.subscribe(() => {
      this.store.set((s) => ({
        ...s,
        location: this.history.location,
      }));
    });
  }
}

export const RouterContext = React.createContext<Router>(undefined!);

export function useRouter<U = RouterState>(select?: (v: RouterState) => U) {
  const router = React.useContext(RouterContext);
  return useStore(router.store, select);
}
