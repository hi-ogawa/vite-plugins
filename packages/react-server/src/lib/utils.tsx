/**
 * @example
 * "/hello/world" =>
 *    [
 *      ["",             ""],
 *      ["/hello",       "hello"],
 *      ["/hello/world", "world"],
 *    ]
 */
export function getPathPrefixes(pathname: string) {
  const keys = normalizePathname(pathname).split("/");
  return keys.map((key, i) => [keys.slice(0, i + 1).join("/"), key] as const);
}

// strip trailing slash (including "/" => "")
export function normalizePathname(pathname: string) {
  return pathname.replaceAll(/\/*$/g, "");
}
