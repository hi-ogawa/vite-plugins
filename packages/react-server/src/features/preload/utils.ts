import { objectMapValues, typedBoolean } from "@hiogawa/utils";
import { matchRouteTree } from "../../utils/tree";
import type { AssetDeps, BuildMetadata } from "./plugin";

export function getRouteAssetsDeps(
  pathname: string,
  manifest: BuildMetadata,
): AssetDeps {
  const result = matchRouteTree(manifest.routeTree, pathname);
  const deps = result.nodes
    .flatMap((n) => [n.value?.page, n.value?.layout, n.value?.error])
    .filter(typedBoolean);
  return mergeAssetDeps(deps);
}

export function mergeAssetDeps(entries: AssetDeps[]): AssetDeps {
  const deps: AssetDeps = { js: [], css: [] };
  return objectMapValues(deps, (_v, k) => entries.flatMap((e) => e[k]));
}
