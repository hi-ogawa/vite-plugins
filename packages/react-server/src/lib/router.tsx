import { objectHas, tinyassert } from "@hiogawa/utils";
import React from "react";
import { type ReactServerErrorContext, createError } from "./error";
import { __global } from "./global";

// cf. https://nextjs.org/docs/app/building-your-application/routing#file-conventions
interface RouteEntry {
  page?: {
    default: React.FC<PageRouteProps>;
  };
  layout?: {
    default: React.FC<LayoutRouteProps>;
  };
  error?: {
    // TODO: warn if no "use client"
    default: React.FC<ErrorRouteProps>;
  };
}

type RouteTreeNode = TreeNode<RouteEntry>;

// generate component tree from glob import such as
//   import.meta.glob("/**/(page|layout).(js|jsx|ts|tsx)"
export function generateRouteTree(globEntries: Record<string, unknown>) {
  const entries: Record<string, RouteEntry> = {};
  for (const [k, v] of Object.entries(globEntries)) {
    const m = k.match(/^(.*)\/(page|layout|error)\.\w*$/);
    tinyassert(m && 1 in m && 2 in m);
    tinyassert(objectHas(v, "default"), `no deafult export found in '${k}'`);
    ((entries[m[1]] ??= {}) as any)[m[2]] = v;
  }

  const flatTree = Object.entries(entries).map(([k, v]) => ({
    keys: k.split("/"),
    value: v,
  }));
  const tree = createTree(flatTree);
  return tree;
}

export type MatchRouteResult = {
  nodes: RouteTreeNode[];
  notFound: boolean;
  params: Record<string, string>;
};

export function matchRoute(
  pathname: string,
  tree: RouteTreeNode,
): MatchRouteResult {
  let node = tree;
  const result: MatchRouteResult = {
    nodes: [],
    notFound: false,
    params: {},
  };
  // strip trailing slash
  const keys = pathname.replaceAll(/\/*$/g, "").split("/");
  for (const key of keys) {
    const next = matchChild(key, node);
    if (next?.child) {
      node = next.child;
      result.nodes.push(node);
      if (next.param) {
        result.params[next.param] = key;
      }
      continue;
    }
    result.notFound = true;
    break;
  }
  return result;
}

// TODO: separate react code in a different file
export function renderMatchRoute(props: RouteProps) {
  const { ErrorBoundary, DefaultRootErrorPage } = __global.clientInternal;

  const nodes = [...props.match.nodes].reverse();

  let acc: React.ReactNode = <ThrowNotFound />;
  if (!props.match.notFound) {
    const Page = nodes[0]?.value?.page?.default;
    if (Page) {
      acc = <Page {...props} />;
    }
  }

  for (const node of nodes) {
    const ErrorPage = node.value?.error?.default;
    if (ErrorPage) {
      // TODO: can we remove extra <div>?
      acc = (
        <ErrorBoundary errorComponent={ErrorPage} url={props.request.url}>
          <div className="error-boundary">{acc}</div>
        </ErrorBoundary>
      );
    }
    const Layout = node.value?.layout?.default;
    if (Layout) {
      acc = <Layout {...props}>{acc}</Layout>;
    }
  }

  acc = (
    <ErrorBoundary
      errorComponent={DefaultRootErrorPage}
      url={props.request.url}
    >
      {acc}
    </ErrorBoundary>
  );

  return acc;
}

const ThrowNotFound: React.FC = () => {
  throw createError({ status: 404 });
};

interface RouteProps {
  request: Request;
  match: MatchRouteResult;
}

export interface PageRouteProps extends RouteProps {}

export interface ErrorRouteProps {
  error: Error;
  serverError?: ReactServerErrorContext;
  reset: () => void;
}

export interface LayoutRouteProps extends React.PropsWithChildren<RouteProps> {}

function matchChild(input: string, node: RouteTreeNode) {
  if (!node.children) {
    return;
  }
  // TODO: sort to dynmaic one come last
  // TODO: catch-all route
  for (const [segment, child] of Object.entries(node.children)) {
    const m = segment.match(/^\[(.*)\]$/);
    if (m) {
      tinyassert(1 in m);
      return { param: m[1], child };
    }
    if (segment === input) {
      return { child };
    }
  }
  return;
}

//
// general tree utils copied from vite-glob-routes
// https://github.com/hi-ogawa/vite-plugins/blob/c2d22f9436ef868fc413f05f243323686a7aa143/packages/vite-glob-routes/src/react-router/route-utils.ts#L15-L22
//

type TreeNode<T> = {
  value?: T;
  children?: Record<string, TreeNode<T>>;
};

function initNode<T>(): TreeNode<T> {
  return {};
}

function createTree<T>(entries: { value: T; keys: string[] }[]): TreeNode<T> {
  const root = initNode<T>();

  for (const e of entries) {
    let node = root;
    for (const key of e.keys) {
      node = (node.children ??= {})[key] ??= initNode();
    }
    node.value = e.value;
  }

  return root;
}
