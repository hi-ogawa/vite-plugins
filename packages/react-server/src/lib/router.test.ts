import { range } from "@hiogawa/utils";
import { describe, expect, it } from "vitest";
import { getPathPrefixes } from "../features/router/utils";
import { generateRouteTree, renderRouteMap } from "./router";

describe(generateRouteTree, () => {
  it("basic", async () => {
    const files = [
      "/layout.tsx",
      "/page.tsx",
      "/other/page.tsx",
      "/test/layout.tsx",
      "/test/page.tsx",
      "/test/other/page.tsx",
      "/test/[dynamic]/page.tsx",
      "/test/[dynamic]/hello/page.tsx",
      "/demo/layout.tsx",
      "/demo/page.tsx",
    ];
    const input = Object.fromEntries(files.map((k) => [k, { default: k }]));
    const tree = generateRouteTree(input);
    expect(tree).toMatchInlineSnapshot(`
      {
        "children": {
          "": {
            "children": {
              "demo": {
                "value": {
                  "layout": {
                    "default": "/demo/layout.tsx",
                  },
                  "page": {
                    "default": "/demo/page.tsx",
                  },
                },
              },
              "other": {
                "value": {
                  "page": {
                    "default": "/other/page.tsx",
                  },
                },
              },
              "test": {
                "children": {
                  "[dynamic]": {
                    "children": {
                      "hello": {
                        "value": {
                          "page": {
                            "default": "/test/[dynamic]/hello/page.tsx",
                          },
                        },
                      },
                    },
                    "value": {
                      "page": {
                        "default": "/test/[dynamic]/page.tsx",
                      },
                    },
                  },
                  "other": {
                    "value": {
                      "page": {
                        "default": "/test/other/page.tsx",
                      },
                    },
                  },
                },
                "value": {
                  "layout": {
                    "default": "/test/layout.tsx",
                  },
                  "page": {
                    "default": "/test/page.tsx",
                  },
                },
              },
            },
            "value": {
              "layout": {
                "default": "/layout.tsx",
              },
              "page": {
                "default": "/page.tsx",
              },
            },
          },
        },
      }
    `);

    function testMatch(pathname: string) {
      const request = { url: "https://test.local" + pathname };
      return renderRouteMap(tree, request as any as Request);
    }

    const testCases = [
      "/",
      "/other",
      "/not-found",
      "/test",
      "/test/other",
      "/test/not-found",
    ];
    for (const i of range(testCases.length)) {
      const testCase = testCases[i]!;
      expect(await testMatch(testCase)).matchSnapshot(`(${i}) "${testCase}"`);
    }
  });
});

describe(getPathPrefixes, () => {
  it("basic", () => {
    expect(getPathPrefixes("/")).toMatchInlineSnapshot(`
      [
        "/",
      ]
    `);
    expect(getPathPrefixes("/hello")).toMatchInlineSnapshot(`
      [
        "/",
        "/hello",
      ]
    `);
    expect(getPathPrefixes("/hello/world")).toMatchInlineSnapshot(`
      [
        "/",
        "/hello",
        "/hello/world",
      ]
    `);
  });
});
