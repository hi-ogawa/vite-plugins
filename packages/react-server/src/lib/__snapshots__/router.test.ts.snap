// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateRouteTree > basic > (0) "/" 1`] = `
{
  "layouts": {
    "/": </layout.tsx
      params={{}}
      request={
        {
          "url": "https://test.local/",
        }
      }
    >
      <LayoutContent
        name="/"
      />
    <//layout.tsx>,
  },
  "pages": {
    "/": </page.tsx
      params={{}}
      request={
        {
          "url": "https://test.local/",
        }
      }
    />,
  },
}
`;

exports[`generateRouteTree > basic > (1) "/other" 1`] = `
{
  "layouts": {
    "/": </layout.tsx
      params={{}}
      request={
        {
          "url": "https://test.local/other",
        }
      }
    >
      <LayoutContent
        name="/"
      />
    <//layout.tsx>,
    "/other": <LayoutContent
      name="/other"
    />,
  },
  "pages": {
    "/other": </other/page.tsx
      params={{}}
      request={
        {
          "url": "https://test.local/other",
        }
      }
    />,
  },
}
`;

exports[`generateRouteTree > basic > (2) "/not-found" 1`] = `
{
  "layouts": {
    "/": </layout.tsx
      params={{}}
      request={
        {
          "url": "https://test.local/not-found",
        }
      }
    >
      <LayoutContent
        name="/"
      />
    <//layout.tsx>,
    "/not-found": <LayoutContent
      name="/not-found"
    />,
  },
  "pages": {
    "/not-found": <ThrowNotFound
      params={{}}
      request={
        {
          "url": "https://test.local/not-found",
        }
      }
    />,
  },
}
`;

exports[`generateRouteTree > basic > (3) "/test" 1`] = `
{
  "layouts": {
    "/": </layout.tsx
      params={{}}
      request={
        {
          "url": "https://test.local/test",
        }
      }
    >
      <LayoutContent
        name="/"
      />
    <//layout.tsx>,
    "/test": </test/layout.tsx
      params={{}}
      request={
        {
          "url": "https://test.local/test",
        }
      }
    >
      <LayoutContent
        name="/test"
      />
    <//test/layout.tsx>,
  },
  "pages": {
    "/test": </test/page.tsx
      params={{}}
      request={
        {
          "url": "https://test.local/test",
        }
      }
    />,
  },
}
`;

exports[`generateRouteTree > basic > (4) "/test/other" 1`] = `
{
  "layouts": {
    "/": </layout.tsx
      params={{}}
      request={
        {
          "url": "https://test.local/test/other",
        }
      }
    >
      <LayoutContent
        name="/"
      />
    <//layout.tsx>,
    "/test": </test/layout.tsx
      params={{}}
      request={
        {
          "url": "https://test.local/test/other",
        }
      }
    >
      <LayoutContent
        name="/test"
      />
    <//test/layout.tsx>,
    "/test/other": <LayoutContent
      name="/test/other"
    />,
  },
  "pages": {
    "/test/other": </test/other/page.tsx
      params={{}}
      request={
        {
          "url": "https://test.local/test/other",
        }
      }
    />,
  },
}
`;

exports[`generateRouteTree > basic > (5) "/test/not-found" 1`] = `
{
  "layouts": {
    "/": </layout.tsx
      params={{}}
      request={
        {
          "url": "https://test.local/test/not-found",
        }
      }
    >
      <LayoutContent
        name="/"
      />
    <//layout.tsx>,
    "/test": </test/layout.tsx
      params={{}}
      request={
        {
          "url": "https://test.local/test/not-found",
        }
      }
    >
      <LayoutContent
        name="/test"
      />
    <//test/layout.tsx>,
    "/test/not-found": <LayoutContent
      name="/test/not-found"
    />,
  },
  "pages": {
    "/test/not-found": </test/[dynamic]/page.tsx
      params={
        {
          "dynamic": "not-found",
        }
      }
      request={
        {
          "url": "https://test.local/test/not-found",
        }
      }
    />,
  },
}
`;
